Tasks/FtpUploadV2/ftpuploadtask.ts

```diff
@@ -23,142 +20,322 @@ export class FtpOptions {
+    if (win) {
+        matchOptions["nocase"] = true;
+    }
+
+    tl.debug("Candidates found for match: " + allFiles.length);
+    for (let i = 0; i < allFiles.length; i++) {
+        tl.debug("file: " + allFiles[i]);
+    }
+
+    // use a set to avoid duplicates
+    let matchingFilesSet: Set<string> = new Set();
+
+    for (let i = 0; i < ftpOptions.filePatterns.length; i++) {
+        let normalizedPattern: string = path.join(
+            ftpOptions.rootFolder,
+            path.normalize(ftpOptions.filePatterns[i])
+        );
+
+        tl.debug("searching for files, pattern: " + normalizedPattern);
+
+        let matched = tl.match(allFiles, normalizedPattern, null, matchOptions);
+        tl.debug("Found total matches: " + matched.length);
+        // ensure each result is only added once
+        for (let j = 0; j < matched.length; j++) {
+            let match = path.normalize(matched[j]);
+            let stats = tl.stats(match);
+            if (!ftpOptions.preservePaths && stats.isDirectory()) {
+                // if not preserving paths, skip all directories
+            } else if (matchingFilesSet.add(match)) {
+                tl.debug(
+                    "adding " + (stats.isFile() ? "file:   " : "folder: ") + match
+                );
+                if (stats.isFile() && ftpOptions.preservePaths) {
+                    // if preservePaths, make sure the parent directory is also included
+                    let parent = path.normalize(path.dirname(match));
+                    if (matchingFilesSet.add(parent)) {
+                        tl.debug("adding folder: " + parent);
+                    }
+                }
+            }
         }
-    });
+    }
+    return Array.from(matchingFilesSet).sort();
+}
 
-    ftpClient.on('close', (hadErr: boolean) => {
-        console.log(tl.loc('Disconnected'));
-        tl.debug('ftp client close, hadErr:' + hadErr);
-    });
+async function run() {
+    tl.setResourcePath(path.join(__dirname, "task.json"));
+
+    const tries = 3;
+    const ftpOptions: FtpOptions = getFtpOptions();
 
-    ftpClient.on('end', () => {
-        tl.debug('ftp client end');
-    })
+    if (!ftpOptions.serverEndpointUrl.protocol) {
+        tl.setResult(tl.TaskResult.Failed, tl.loc("FTPNoProtocolSpecified"));
+    }
+    if (!ftpOptions.serverEndpointUrl.hostname) {
+        tl.setResult(tl.TaskResult.Failed, tl.loc("FTPNoHostSpecified"));
+    }
 
-    ftpClient.on('error', (err) => {
-        tl.debug('ftp client error, err: ' + err);
-        if (!uploadSuccessful) {
-            // once all files are successfully uploaded, a subsequent error should not fail the task
-            failTask(err);
+    const files: string[] = findFiles(ftpOptions);
+    let ftpClient = await getFtpClient(ftpOptions);
+    const tracker = new ProgressTracking(ftpOptions, files.length + 1); // add one extra for the root directory
+
+    let retryWithNewClient = async (task: () => {}, retry: number) => {
```

Comment:
It seems to me that `retryWithNewClient` would be better as a class.  Probably fine for this use, but `tries` and a whole lot of the `awaits` below could probably be streamlined a bit if it were something more like:
```
const clientRunner = new ClientRunner(MAX_TRIES);
await clientRunner.run(async (ftp) => ftp.ensureDir());
```

Reply:
You need a reference to the ftp client in order to close the connection during retry.  That's why it's a local function.  In the case of a class (your second line), you got to wrap the ftp client object with some other wrappers to properly pass it around as you need to initiate a new ftp client if connection dropps.  Too much trouble for something minor.

Reply:
You could be right.  Here's a more thorough explanation of what I was talking about:
```
class ClientRunner {
  private ftpClient: FtpClient;
  constructor(options: FtpOptions, maxRetries: number) {
    ftpClient = null;
    // save other values
  }

  public async run(t: async (ftp: FtpClient) => void) {
    if (!this.ftpClient && this.currentRetries < this.maxRetries) {
      this.ftpClient = await getFtpClient(this.ftpOptions);
    }
    // your retryWithNewClient code
    //  instead of 'await task()'
    await task(this.ftpClient);
  }
}
```
Seems cleaner to me.  YMMV.

Reply:
Yeah, this is wrapping the ftp client.  :)  Also, we may not be able to define retry in the constructor, as each run has different costs - for some call, the client library may make hundreds calls (like delete folder, it needs to recursively remove every file first), and server can drop connection at anytime, we may want to have a higher retry count on those operations.  So the `run` method need to take the retry count anyway, then it looks almost the same as a local method.  :D